src/viam/sdk/common/proto_type.hpp

```diff
@@ -27,64 +27,82 @@ struct get_helper {
     }
 };
 
-template <>
-struct get_helper<AttributeMap> {
-    template <typename V>
-    static AttributeMap get(V& v) {
-        auto* const result = boost::get<AttributeMap>(&v);
-        return result ? *result : nullptr;
-    }
-
-    template <typename V>
-    static std::shared_ptr<const AttributeMap::element_type> get(const V& v) {
-        auto* const result = boost::get<AttributeMap>(&v);
-        return result ? *result : nullptr;
-    }
-};
-
-}  // namespace prototype_details
+}  // namespace value_type_details
 
-class ProtoType {
+class AttributeMap {
    public:
-    ProtoType() {
-        proto_type_ = boost::blank();
-    }
-
-    explicit ProtoType(bool b) : proto_type_(std::move(b)) {}
-    explicit ProtoType(std::string s) : proto_type_(std::move(s)) {}
-    explicit ProtoType(const char* c) : proto_type_(std::string(c)) {}
-    explicit ProtoType(int i) : proto_type_(std::move(i)) {}
-    explicit ProtoType(double d) : proto_type_(std::move(d)) {}
-    explicit ProtoType(AttributeMap m) : proto_type_(std::move(m)) {}
-    explicit ProtoType(std::vector<std::shared_ptr<ProtoType>> v) : proto_type_(std::move(v)) {}
-    explicit ProtoType(const google::protobuf::Value& value);
-
-    google::protobuf::Value proto_value();
-    friend bool operator==(const ProtoType& lhs, const ProtoType& rhs);
-
-    template <typename T>
-    auto get() {
-        return prototype_details::get_helper<T>::get(proto_type_);
-    }
-
-    template <typename T>
-    auto get() const {
-        return prototype_details::get_helper<T>::get(proto_type_);
-    }
+    class value_type;
+
+    using map_type = std::unordered_map<std::string, value_type>;
+
+    class value_type {
+       public:
+        template <typename T>
+        auto get() {
+            return value_type_details::get_helper<T>::get(value_);
+        }
+
+        template <typename T>
+        auto get() const {
+            return value_type_details::get_helper<T>::get(value_);
+        }
+
+        using base_value_types_ =
+            boost::variant<boost::blank, std::string, const char*, int, float, double, bool>;
+        using value_types = boost::make_recursive_variant<
+            base_value_types_,
+            std::unordered_map<std::string, boost::recursive_variant_>,
+            std::vector<boost::recursive_variant_>>::type;
+
+        value_type();
+        value_type(std::string s);
+        value_type(const char* c);
+        value_type(int i);
+        value_type(float f);
+        value_type(double d);
+        value_type(bool b);
+        value_type(std::vector<value_type> v);
+        value_type(map_type);
+        value_type(google::protobuf::Value value);
```

Comment:
For at least these types, copying is pretty expensive. I recommend providing both a `const&` and a `&&` overload.

Alternatively, you could write a universally forwarding constructor:

```
template<typename T>
value_type(T&& t)
```

However, that tends to be too greedy without some SFINAE to make sure it doesn't get invoked when it shouldn't.